keanu-project/src/main/java/io/improbable/keanu/network/ProtobufSaver.java

```diff
@@ -16,28 +16,55 @@
 import java.io.IOException;
 import java.io.OutputStream;
 import java.lang.reflect.Method;
+import java.util.Map;
 import java.util.function.Consumer;
 
 public class ProtobufSaver implements NetworkSaver {
     private final BayesianNetwork net;
-    private KeanuSavedBayesNet.BayesianNetwork.Builder bayesNetBuilder = null;
+    private KeanuSavedBayesNet.Model.Builder modelBuilder = null;
 
     public ProtobufSaver(BayesianNetwork net) {
         this.net = net;
     }
 
+    public KeanuSavedBayesNet.Model getModel(boolean withSavedValues, Map<String, String> metadata) {
+        createProtobufModel(withSavedValues, metadata);
+        return modelBuilder.build();
+    }
+
+    @Override
+    public void save(OutputStream output, boolean saveValues, Map<String, String> metadata) throws IOException {
+        createProtobufModel(saveValues, metadata);
+        modelBuilder.build().writeTo(output);
+        modelBuilder = null;
+    }
+
     @Override
     public void save(OutputStream output, boolean saveValues) throws IOException {
-        bayesNetBuilder = KeanuSavedBayesNet.BayesianNetwork.newBuilder();
+        createProtobufModel(saveValues, null);
+        modelBuilder.build().writeTo(output);
+        modelBuilder = null;
+    }
+
+    private void createProtobufModel(boolean saveValues, Map<String, String> metadata) {
+        modelBuilder = KeanuSavedBayesNet.Model.newBuilder();
 
         net.save(this);
 
         if (saveValues) {
             net.saveValues(this);
         }
+        saveMetadata(metadata);
+    }
 
-        bayesNetBuilder.build().writeTo(output);
-        bayesNetBuilder = null;
+    private void saveMetadata(Map<String, String> metadata) {
+        if (metadata != null) {
+            KeanuSavedBayesNet.Metadata.Builder metadataBuilder = KeanuSavedBayesNet.Metadata.newBuilder();
+            for (Map.Entry<String, String> entry : metadata.entrySet()) {
+                metadataBuilder.putMetadataInfo(entry.getKey(), entry.getValue());
+            }
+            modelBuilder.setMetadata(metadataBuilder).build();
```

Comment:
I think your .build() is in the wrong place here?  Don't you want to put that on the metadataBuilder?

Reply:
Very true :) I removed the build altogether, it still works.